### PR TITLE
Conditional rewriting

### PR DESCRIPTION
**Pull request by francoisthire**
This PR is a proposition to introduce conditional rewriting in Dedukti with the syntax proposed in #73 . Even if the code is quite ugly right now, it is already possible to play around with conditional rewriting.

### CODE REVIEW

File: commands/dkcheck.ml [hunk 1]
```diff
@@ -156,8 +156,8 @@ let _ =
       , Arg.String Confluence.set_cmd
       , "CMD Set the external confluence checker command to CMD" )
     ; ( "-nl"
-      , Arg.Set Rule.allow_non_linear
-      , " Allow non left-linear rewriting rules" )
+      , Arg.Set Rule.allow_conditional
+      , " Allow non conditional rewriting rules" )
```
Comment (rafoo) [inline commands/dkcheck.ml]: Probably a typo here: we always want non-conditional rewrite rules, the question is wether or not we want conditional ones.

File: kernel/rule.ml [hunk 2]
```diff
@@ -17,28 +17,30 @@ type wf_pattern =
...
 type rule_name = Delta of name | Gamma of bool * name
 
+type condition = {left:term; right:term}
+
 type 'a rule =
   {
     name: rule_name;
     ctx: 'a;
     pat: pattern;
+    cond: condition option;
```
Comment (rafoo) [inline kernel/rule.ml]: From what I understand from the rest of the code, we could have a list of conditions (interrpreted as their conjunction) here.

File: kernel/rule.mli [hunk 3]
```diff
@@ -42,13 +41,15 @@ type rule_name =
   (** Rules of lambda pi modulo. The first parameter indicates whether
       the name of the rule has been given by the user. *)
 
+type condition = {left:term; right:term}
```
Comment (rafoo) [inline kernel/rule.mli]: Would it be hard to add a case to negate a conversion check? The constraint t1 != t2 would mean that the head-normal forms of t1 and t2 tell us that no substitution can make t1 and t2 convertible.

File: kernel/rule.ml (hunk 2 above)
Comment (francoisthire) [reply kernel/rule.ml]: This is done in my last commit.

File: kernel/rule.mli (hunk 3 above)
Comment (francoisthire) [reply kernel/rule.mli]: Guards are not used during the type checking of a rule. But I have implemented a negation condition that is true if and only if the two terms are NOT convertible when Dedukti tries to apply the rule.

File: kernel/rule.ml (hunk 2 above)
Comment (rafoo) [reply kernel/rule.ml]: Nice!

**Comment by francoisthire** [comment]
True, while I was implementing this, I was thinking there is a problem but
I did not look further. Maybe for the moment, we can do without the
negation? Introducing negation should be IMHO in another PR.

2018-04-19 11:20 GMT+02:00 Raphaël Cauderlier <notifications@github.com>:

**Comment by francoisthire** [comment]
We had a little chat about this with @rlepigre and @GuillaumeGen and the conclusion is that this PR does not change the expressivity of Dedukti. It only simplifies encodings in Dedukti.

**Comment by fblanqui** [comment]
What do you mean by "expressivity"?


Le 15/09/2018 à 16:02, fthire a écrit :

**Comment by francoisthire** [comment]
@rafoo The last commit removed the negation things. I can spend some times on this PR if you think that this feature can be used for the REC competition.

**Comment by francoisthire** [comment]
Too old I close.

**Pull request closed without merging**